util/design/secded_gen.py

```diff
@@ -64,75 +59,74 @@ def calc_fanin(width, codes):
-    """
-    outstr = ""
-    line = prefix
-    prepend_len = len(prefix)
-    cnt = start_cnt
-    first = True
-    for j in range(k):
-        temp_str = ""
-        if cur_m in codes[j]:
-            if not first:
-                temp_str += " ^"
-            if first:
-                first = False
-            temp_str += " in[%d]" % (j)
-            temp_len = len(temp_str)
-
-            if len(line) + temp_len > max_width:
-                outstr += line + "\n"
-                line = ' ' * (prepend_len - first_indent) + temp_str
-            else:
-                line += temp_str
-    outstr += line + ";\n"
-    return outstr
+def calc_bitmasks(k, m, codes, dec):
+    # Transform fanin indices into bitmask.
+    fanin_masks = [0] * m
+    for i, c in enumerate(codes):
+        for j in c:
+            fanin_masks[j] += 1 << i
+    # For decode ops, include ECC bit position.
+    if dec:
+        for j in range(m):
+            fanin_masks[j] += 1 << (k + j)
+    return fanin_masks
 
 
 def print_enc(n, k, m, codes):
-    outstr = ""
-    for i in range(k):
-        outstr += "  assign out[%d] = in[%d] ;\n" % (i, i)
-
-    for i in range(m):
-        # Print parity computation
-        outstr += print_comb(n, k, m, i, codes, 0, 100,
-                             "  assign out[%d] =" % (i + k), 2)
+    outstr = "  always_comb begin : p_encode\n"
+    outstr += "    out = {}'(in);\n".format(n)
+    format_str = "    out[{}] = ^(out & " + str(n) + "'h{:0" + str(
+        (n + 3) // 4) + "X});\n"
+    # Print parity computation
+    for j, mask in enumerate(calc_bitmasks(k, m, codes, False)):
+        outstr += format_str.format(j + k, mask)
+    outstr += "  end\n"
     return outstr
 
 
 def calc_syndrome(code):
-    log.info("in syncrome {}".format(code))
+    log.info("in syndrome {}".format(code))
     return sum(map((lambda x: 2**x), code))
 
+
+def print_dec(n, k, m, codes, codetype):
+    outstr = ""
+    outstr += "  logic single_error;\n"
+    outstr += "\n"
+    outstr += "  // Syndrome calculation\n"
+    format_str = "  assign syndrome_o[{}] = ^(in & " + str(n) + "'h{:0" + str(
+        (n + 3) // 4) + "X});\n"
+    # Print syndrome computation
+    for j, mask in enumerate(calc_bitmasks(k, m, codes, True)):
+        outstr += format_str.format(j, mask)
+    outstr += "\n"
+    outstr += "  // Corrected output calculation\n"
+    for i in range(k):
+        outstr += "  assign d_o[%d] = (syndrome_o == %d'h%x) ^ in[%d];\n" % (
+            i, m, calc_syndrome(codes[i]), i)
+    outstr += "\n"
+    outstr += "  // err_o calc. bit0: single error, bit1: double error\n"
+    # The Hsiao and Hamming syndromes are interpreted slightly differently.
+    if codetype == "hamming":
+        outstr += "  assign err_o[0] = syndrome_o[%d];\n" % (m - 1)
+        outstr += "  assign err_o[1] = |syndrome_o[%d:0] & ~syndrome_o[%d];\n" % (
+            m - 2, m - 1)
+    else:
+        outstr += "  assign single_error = ^syndrome_o;\n"
```

Comment:
is there a reference you can recommend?  I want to make sure i actually understand why the hamming syndrome is treated differently.

As far as I can tell, since the msb syndrome bit is the xor of everything, I guess it's effectively the final parity check that tells if you if there's a bit-flip somewhere? So does that imply that any odd number of bit-flips would all get registered as a "single" error, and all even ones would be registered as double error?

Reply:
Yes exactly. 
The final parity bit behaves differently than in the Hsiao code, since it also includes all other ECC bits.

This paper here explains Hamming and Hsiao codes in sections 3. and 4. and points out the differences:
http://ysu.am/files/11-1549527438-.pdf

Reply:
I think I am going to link this paper somewhere in the script for reference...